### ISSUE TITLE
Drop support for old PHP version?

### ISSUE THREAD

**Issue by Schnoop** [friendica/friendica#12210]
Are there any plans to drop support for old PHP versions? PHP 7.3 is out of any support and PHP 7.4 security support will end on 28 Nov 2022. Maybe it is a good point to move forward and cut those old stuff and invest in the overall code quality.

**Comment by MrPetovan** [comment]
If only!

The sad truth is that both developers and many admins can be stuck with PHP 7 depending on their server OS. In my personal case, I'm running the Devuan distribution (Debian without systemd) and it usually is behind Debian, which is usually late updating to new PHP versions.

When Devuan releases their next version including PHP 8.1, I believe we'll be in good shape to drop support for PHP 7 but the time hasn't come yet.

**Comment by tobiasd** [comment]
We are orientating the lower supported version of Friendica on Debian stable, give some time so that old-stable can update to stable (and derivates like Devuan update as well as MrPetovan pointed out). At the moment this would be PHP7.4.

**Comment by Art4** [comment]
Today (2026-02-23) 
- [Debian stable (trixie) comes with PHP 8.4](https://packages.debian.org/stable/php)
- [Debian oldstable (bookworm) comes with PHP 8.2](https://packages.debian.org/bookworm/php)

I would argue that it would be safe enough if we drop support for PHP 7.4, 8.0 and 8.1 and require at least PHP 8.2.

This would allow us to upgrade PHPUnit from v9, skipping v10 (that has some downsides, like no built-in way to test user deprecations) and upgrade directly to [PHPUnit 11](https://phpunit.de/announcements/phpunit-11.html),

**Comment by annando** [comment]
I wouldn't want to drop also 8.0 and 8.1 if not really needed. Which versions come with which advantages?

**Comment by Art4** [comment]
This is a (highly opinionated) list of features that has real effects on my daily work with PHP and imho on the work on Friendica. The data is based on [php.net](https://www.php.net/supported-versions.php) and [php.watch](https://php.watch/versions).

## tl;dr

We should update to PHP 8.2 ✔️ , or at least PHP 8.1 ⚠️ , but not PHP 8.0 ❌ .

Only with PHP 8.2, we finally have **full type support** for `true`, `false`, `never`, and `mixed` without constantly having to think about which types are supported and which are not. And we can **update our test code** together with the production code.

## [PHP 8.0](https://www.php.net/manual/en/migration80.php)

- [named parameters](https://www.php.net/manual/en/functions.arguments.php#functions.named-arguments)
- [attributes](https://www.php.net/manual/en/language.attributes.php)
- [constructor properties](https://www.php.net/manual/en/language.oop5.decon.php#language.oop5.decon.constructor.promotion)
- [union types](https://www.php.net/manual/en/language.types.declarations.php#language.types.declarations.composite.union)
- allow trailing comma in parameter lists
- [new `mixed` type](https://www.php.net/manual/en/language.types.mixed.php)
- add `false` as special union type
- [str_contains()](https://www.php.net/manual/en/function.str-contains.php), [str_starts_with()](https://www.php.net/manual/en/function.str-starts-with.php) and [str_ends_with()](https://www.php.net/manual/en/function.str-ends-with.php)
- [resource to object migrations](https://php.watch/articles/resource-object) like the new [CurlHandle](https://www.php.net/manual/en/class.curlhandle.php) class (and drop of `resource` for curl functions)

❌ We cannot update PHPUnit 9.
❌ PHP 8.0 is end-of-life since Nov 2023
❌ [Debian oldstable (bookworm) comes with PHP 8.2](https://packages.debian.org/bookworm/php)

## [PHP 8.1](https://www.php.net/manual/en/migration81.php)

- Support for [Enums](https://www.php.net/manual/en/language.enumerations.php) has been added.
- [first class callable syntax](https://www.php.net/manual/en/functions.first_class_callable_syntax.php)
- Support for [intersection types](https://www.php.net/manual/en/language.types.type-system.php#language.types.type-system.composite.intersection) has been added.
- [Readonly Properties](https://php.watch/versions/8.1/readonly)
- [new `never` type](https://www.php.net/manual/en/language.types.never.php)

⚠️ We can update to PHPUnit 10.
❌ PHP 8.1 is end-of-live since Dec 2025
❌ [Debian oldstable (bookworm) comes with PHP 8.2](https://packages.debian.org/bookworm/php)

## [PHP 8.2](https://www.php.net/manual/en/migration82.php)

- new [`#[\SensitiveParameter]` attribute](https://php.watch/versions/8.2#sensitive-parameter)
- new `true` type
- It is now possible to use [null](https://www.php.net/manual/en/language.types.null.php) and [false](https://www.php.net/manual/en/language.types.singleton.php) as stand-alone types.
- It is now possible to combine intersection and union types
- [readonly classes](https://www.php.net/manual/en/language.oop5.basic.php#language.oop5.basic.class.readonly)

✔️ We can update to PHPUnit 11.
✔️ PHP 8.2 gets security updates until Dec 2026
✔️ [Debian oldstable (bookworm) comes with PHP 8.2](https://packages.debian.org/bookworm/php)

## About PHPUnit

We use PHPUnit for our test code. When we update the PHP version, we will want to update our production code. We have to do the same with our test code.

✔️ With PHP 8.2, we can update to PHPUnit 11. This brings various features such as testing deprecations and other warnings.

⚠️ With PHP 8.1, we can at least update to PHPUnit 10, but this comes with some hard breaking changes (new event system without backward compatibility), which makes tool support difficult (which is why even the [yoast/PHPUnit-Polyfills project](https://github.com/Yoast/PHPUnit-Polyfills) has skipped support PHPUnit 10).

❌ With PHP 8.0, we have no or very few improvements to our test suite, as we have to stick with PHPUnit 9. We cannot switch from annotations to attributes in the test code, which will have to be done at some point later instead of updating it directly with the production code.

**Comment by MrPetovan** [comment]
Thank you for taking the time for a comparison. However, the choice of supporting older PHP versions has never been about capabilities, but about adoption. If there are a significant number of Friendica nodes running PHP 8.0, we simply can't raise the minimum PHP version, no matter what goodies newer versions bring.

The detail about Debian oldstable is welcome as it is a valid factor in this context.

**Comment by annando** [comment]
@MrPetovan by now it seems as if no one is using a PHP version lower than 8.2: https://forum.friendi.ca/display/ec054ce7-7069-9d4f-74e1-9f4963731566

**Comment by MrPetovan** [comment]
Seems like we do have a consensus for PHP 8.2 as the baseline!
